### DISCUSSION TITLE
S3 log storage writes - CreateIfNotExists guard

### DISCUSSION THREAD

**Discussion by xianwill** [delta-io/delta-rs#89] [Ideas]
S3 does not have an API that supports the "Create if not exists" semantics that the other storage backends have. This discussion is to work out what a wrapper on top of S3 should look like to provide these semantics when writing delta log entries.

Ultimately, we need the `put_obj` [implementation](https://github.com/delta-io/delta-rs/blob/4170d34c90da21ee406c0aa0a15dc5bf09eba011/rust/src/storage/s3.rs#L184) within the s3 backend to fail with `StorageError:AlreadyExists` as [the fs backend](https://github.com/delta-io/delta-rs/blob/4170d34c90da21ee406c0aa0a15dc5bf09eba011/rust/src/storage/file.rs#L78) does iff a log entry with the same version has already been written.

DynamoDb is frequently mentioned as a metastore on top of S3 to provide this type of API. delta-rs should provide functionality to:

* create the required DynamoDb schema for guarding log entry writes to S3
* implement a reliable protocol within the s3 backend put_obj method that leverages DynamoDb to guard S3 writes

This discussion is intended to work out the schema and protocol to include "Create if not exists" semantics in the put_obj implementation of the S3 backend.

---

To kick the discussion off with a naive approach:

Consider a DynamoDb schema with only a `path` attribute. A put_obj implementation may look like:

* Write path to DynamoDb table with a put condition to ensure path does not already exist
* On success - write delta log file to s3
* On fail - continue with optimistic concurrency retry loop

Already, we have some problems. If there is a system failure between DynamoDb guard write and delta log file write for version _n_, the delta log file for version _n_ will not be written. The [delta protocol](https://github.com/delta-io/delta/blob/master/PROTOCOL.md#delta-table-specification) requires contiguous, monotonically-increasing integers for version so this must be resolved before other log writers can continue the optimistic concurrency loop with any chance of success.

We _could_ implement a distributed lock w/ timeout around the DynamoDb write and S3 log entry file write, but now we have a distributed lock inside of an optimistic concurrency loop which smells pretty funny to me. Perhaps this is tolerable since the goal is to bring the S3 semantics up to speed with other storage backends.

@houqp + @rtyler - your turn now.

**Comment by xianwill** [comment]
In regards to wrapping the guard and S3 write in a distributed lock, I came across https://crates.io/crates/dynalock recently which may be worth taking a dependency on.

**Comment by houqp** [comment]
The problem you mentioned exists in our current filesystem backend as well, for example: a writer could crash right after it created the new log entry, but before fully writes out the log content. When this happens, we would end up with a corrupted log entry. The core of the problem here is transaction commit semantic is modeled by creation of the file without taking into account of the file content. Same analysis can be applied to the failure case you mentioned with dynamodb, where putting/reserving a key in dynamodb only indicates "creation" of the file, not a file with fully flushed content.

I think it would be more fitting to change the `Create if not exists` semantic requirement to `Atomic rename/rename if not exists` semantic. Take filesystem backend as an example again. What we should do is to write out new transaction log into a temporary path. After the write is fully flushed and persisted, we call a rename system call to rename the temp file to the next available commit version. `Atomic rename` should guarantee a file already exists error to be thrown if another writer has already claimed this version. This will make sure we never commit corrupted log entries.

Back to S3 + Dynamodb, my original suggestion of leveraging S3's new consistent read capability won't fly as you already pointed out. It looks like to be 100% safe, we would still need to use dyanmodb for listing so we can get that `Atomic rename` support.

Here is **_one_** way to do it:

In our commit loop:

1. Write out log entry to a temp and unique S3 path 
1. Atomically put a new key in dynamodb with key set to new transaction version and value set to the above temp S3 path
    - If key already exists, apply all new log entries based on a new dynamodb key scan (this is where we use dynamodb for listing, not s3). We can also store latest version in dynamodb in a dynamodb transaction to avoid the table scan.

To make it backwards compatible with readers using vanilla S3 backends, we can replica logs from dynamodb into S3 in a dedicated single threaded replicator/writer:

1. Go through dyanmodb table to find all unreplicated log entries
1. One by one copy unreplicated log entry into `_delta_log` folder with name set to the corresponding transaction version conforming to the delta spec
   * Once a log entry is fully copied to S3 with log version as the name, mark this particular log entry in dynamodb as replicated. We can mark it by setting key value to a null or use other metadata fields.
   * Delete the old random S3 log entry

**Comment by houqp** [comment]
Another way to do it is using the distributed lock design you mentioned to serialize the write:

1. write out log entry to a temp and unique s3 path
2. grab the distributed lock
3. check if next version in s3 log directory already exists (here we are leveraging s3's new consistent read/listing capability)
4. copy temp s3 log entry into log directory with version as the object name
5. release distributed lock

Notice in this design, we only need to use dynamodb for locking. We don't need to use it to implement `create if not exists` semantic because the lock already guarantees only one writer will be able to reserve s3 path for the next version. Effectively, we are doing `Atomic rename` by using a distributed lock to safe guard the s3 rename operation. S3 doesn't support native object rename though, so it will be implemented as a list + copy + delete. This will take extra time if the S3 object has a large size, which I don't think is the case for log entries.

This design is a simpler compared to the one that uses dynamodb for listing, at the cost of slower commit time and the need to handle lock expiration edge-case.

**Comment by houqp** [comment]
There are a lot of of interesting discussions around this topic in https://github.com/delta-io/delta/issues/41 as well, which covered a lot of the points we discussed here.

**Comment by xianwill** [comment]
Fantastic suggestions all around @houqp. I am all on board for implementing "atomic rename" instead of "create if not exists".

Since the use case I am primarily interested in begins with headless multi-process streaming writers appending to the same table and competing over the next version, I personally think I prefer approach 2 you mention above. I think approach 1 would land me right back into a spot where I would either need a dedicated replication process or another not-yet-mentioned lock to let separate threads in each process compete over _which_ process is doing the replication. Approach 2 does not have this same development overhead as far as i can tell.

**Reply by houqp** [reply]
Yeah, approach 2 is mostly trading performance and liveness for less complexity. I think putting a lock around S3 copy shouldn't be too bad performance wise because commit log entries are usually small in size. We need to have a solution for the following edge-case though:

1. writer A writes obtains transaction lock, claims log version N
2. writer A starts log entry S3 copy
3. writer A temporarily gets stuck due to bugs or network issue
4. lock times out
5. writer B grabs log and claims the same log version N
6. writer B completes S3 copy, releases lock. at this point, transaction has been committed.
7. writer A back in business, assuming that it still holds the lock, continues to finish up the stale S3 copy.

The system should be able to handle lock timeouts caused by hard crashes without issue. But timeouts caused by this kind of intermittent issues could be tricky to handle.

With regards to approach 1, I would refer us going with a single dedicated replicator to keep it simple since the log copy should be quick and easy. As you already mentioned, distributing it across processes brings us back to the lock liveness issue while approach 1's biggest selling point is lockless.

Lastly, what I mentioned so far are just two of the many possible solutions we can explore :) There might be more elegant solutions out there.

**Reply by xianwill** [reply]
Re: the edge case - yep. I was thinking about the same thing. My thought is that DynamoDb interactions either need to use DynamoDb transactions, or otherwise include a put-condition that ensures that the held lock has not timed out and handle the failure by resuming the next iteration of the optimistic concurrency loop if the transaction or put-condition fails for lock timeout.

**Reply by houqp** [reply]
The failure scenario I mentioned in the previous comment is not a hard failure, but rather an unexpected pause. From the first lock holder's point of view, it doesn't know it has been paused for a long time, so after resumed, it will just keep proceeding in the critical section assuming it still owns the expired lock.

https://martin.kleppmann.com/2016/02/08/how-to-do-distributed-locking.html has a good demonstration of the brain split problem that I constructed in my previous reply. https://hazelcast.com/blog/long-live-distributed-locks/ is another implementation focused post worth reading.

Usually, this is solved using a fencing token, which is basically a monotonically increasing counter that get issued on every lock operation. However, to effectively use a fencing token, we need all external systems involved within the critical section to support conditional writes, which is not the case for S3. The `amazon-dynamodb-lock-client` project had a similar discussion around this subject: https://github.com/awslabs/amazon-dynamodb-lock-client/issues/1#issuecomment-323494652.

One thing we can do here is to adopt a 2PC design for the S3 rename. When a writer acquires a lock to perform the rename, it also atomically record this intention in dynamodb by writing out the temp object key. We can think of this as the prepare stage in 2PC. Once the prepare stage is done, we can consider the rename has been completed logically. Once the S3 write has been completed, the writer will perform a final commit in dynamodb to mark the prepared stage as completed.

Now consider the following failure scenarios:

* The first writer crashed after the prepare stage without finishing up the S3 write. The next writer that acquires the expired lock will try to complete the rename by noticing an uncommitted prepare action
* The writer crashed after completed the S3 write, but before committing the rename. The next writer that acquires the expired lock will try to finish up the commit
* The first writer paused after the prepare stage while writing to S3. When the next writer grabs the expired lock, it will try to complete the same S3 rename operation by coping the same temp S3 object to the same prepared transaction log entry. The S3 write is idempotent.
* The first writer paused after completing the S3 write but before committing the prepare stage, it will simply compete with the next writer that acquires the expired lock to commit the same prepared stage. This dynmodb write is also idempotent.

**Comment by houqp** [comment]
@mosyp @xianwill here is an expanded potential atomic rename design with fencing token.

## Before the optimistic commit loop

Write log entries to a temp s3 location (`tmp_key` = prefix + uuid).

## At the beginning of the optimistic commit loop:

Compute the next version from cached delta table version in memory. It is required to compute the version outside of the lock for non-append only writes. For append only writes, we could optimize further to get the latest version from dynanmodb lock.

## Atomic rename

1. Acquire dynamodb lock with `fencing_token`
    * At this point, we have entered the critical section. Keep in mind that because our lock has expiration, it can expire between any atomic operations within the critical section. So the lock itself doesn't provide correctness/safety guarantee. 
1. Repair prepared commits if any. These prepared commits are idempotent/static, so it's safe for multiple writers to repair the same commit concurrently
    1.  for each prepared commit `(tmp_key, delta_version, fencing_token)` in dynamodb:
        1. perform s3 write from `tmp_key` to `delta_version`
            * if `tmp_key` is already deleted by another writer, we go straight to mark the prepared commit as committed
        1. perform s3 delete for `tmp_key`
        1. mark prepared commit as committed in dynamodb
1. Use S3 HEAD API to check if `delta_version` already exists in S3. If yes, return Object already exists error back to caller.
    * Lock could expire after the S3 version conflict check, that means multiple writers could pass this check with the same version. However, the subsequent dyanmodb conditional put with fencing token will guarantee only one of them goes through.
1. Prepare a commit with the tuple `(tmp_key, delta_version, fencing_token)`. In the dynamodb conditional put, we make sure both `fencing_token` and `delta_version` are greater than all existing prepared commit dynamodb items. If conditional PUT failed, return object already exists error back to the caller.
    * Once the dynamodb PUT succeeds, the rename operation has been logically committed in the system and there is no going back. In other words, the prepared rename will eventual become visible to all readers.
    * The current writer can crash or pause from now on and it doesn't matter because other writers will help finish up the prepared commit.
1. "repair" newly prepared commit

**Reply by mosyp** [reply]
> Compute the next version from cached delta table version in memory. It is required to compute the version outside of the lock for non-append only writes

Could you elaborate on why we should compute next version outside of a lock? And what's the difference between append/non-append writes for such behaviour?

**Reply by houqp** [reply]
Append only table only adds new data, so it won't cause transaction conflicts. That means if a new table version has been committed right before the current writer tries to write out the new transaction, it won't need to read the latest committed transactions and reconstruct table states to check for transaction conflicts. It should be safe for it to append files using the next available versions. However, after the transaction is committed, we still need to read new commits committed by other writers to refresh local table states.

For other types of writes, it could lead to transaction conflicts. For example:

* transaction A merges file1 and file2 into file3, which expands to the following actions:
    * add file3
    * delete file1 and file 2
* transaction B deletes file1

Let's say both transactions are being committed with the same version concurrently from different writers with version `N`. If transaction A goes through first, transaction B cannot be committed as `N+1` as is anymore. Because in version `N`, file1 has already been deleted (but its content is still in file3). Thus it will need to catch up on version `N` to reconstruct local table state, then change the transaction to the following steps:

1. read file3 and delete file1 content
2. write purged content to file4 and add to delta table
3. delete file3

See how a single action transaction gets rewritten into two actions with add and delete. Some of these transaction conflict resolutions could read/write a lot of data and take a long time, so we shouldn't do it while holding the lock.

This means we don't get much value from incrementing the version from within the lock because the lock will need to be released for conflict resolution. During this process, we will be updating local table versions as well, so might as well handle the next version increment there.

**Reply by mosyp** [reply]
@houqp So the reason I went with locking outside of s3 backend in https://github.com/delta-io/delta-rs/pull/212 is about this step 
> 3. Repair prepared commits if any. 

This should be done within the lock for sure but its behaviour does not "fit" into `rename` function, at least semantically. But having lock outside of s3 `rename` will allow to add more code into _locked_ section. 

Anyways, I'm happy to go as is and put lock into `rename` function in s3 and then decided on how to approach with repairing commits when we'll get a clearer picture

**Reply by xianwill** [reply]
@mosyp Took me a bit to wrap my head around this, but I think this still works fully encapsulated within the S3 backend. A paused worker that previously held a lock will have recorded its intent (as tmp_key and delta_version) while holding the lock. So the S3 backend layer in all worker contexts has full visibility to that intent. As long as the first thing the S3 backend does is try to repair a prior existing commit by reading the intent recorded in the expired lock - this should be fine. Most likely the S3 backend will have to return an "object exists" error to the caller which triggered the repair operation (by trying to commit a different transaction), but that is fine because it will be handled by the optimistic concurrency loop. The writer will go look up the next available version and immediately try another commit. From the writer side, there just is not a semantic difference between a commit that fails because of a repair and a commit that fails from a separate active writer competing over the same version. Fortunately, I don't think we need that level of semantics at the writer layer.

**Reply by houqp** [reply]
> This should be done within the lock for sure but its behaviour does not "fit" into rename function, at least semantically.

We are dealing with two types of commits here, so for sake of clarity, I am going to refer the commits from optimistic transaction commit loop as delta commits and commits from S3 atomic rename as rename commits.

Could you expand a little bit more on why you think the repair behavior doesn't fit into the rename function? Are you concerned about the complexity or is there any technical reason that requires us handling it outside of rename? My original idea was keeping the rename commit, repair actions as implementation details for atomic rename method. Following @xianwill 's line of thought, from the optimistic commit loop point of view, all it cares is whether the rename will fail with already exists error or not, and if so, retry delta commit with a new version.

One minor thing to correct in @xianwill 's comment is I don't think repair will ever cause the delta commit to fail, unless there is a network error or Dynamodb/S3 is down. Prepared rename commit entries are safe to be repaired by multiple writers concurrently by design. One steps that would result in object already exists are the subsequent S3 HEAD (step 3) and dynamodb conditional PUT call (step 4).

**Reply by xianwill** [reply]
@houqp - I may have spoken unclearly. The case I was referring to when I mentioned 

> Most likely the S3 backend will have to return an "object exists" error to the caller

is that - assuming a lock was acquired by worker _A_ with intent _i1_ recorded in lock _L1_ to commit `tmp_key` _k1_ as `delta_version` _v1_, and that the lock expired due to worker pause, when worker _B_ tries to prepare a different commit with intent _i2_, the first operation performed by the S3 backend should be to try to repair failed intents. So rather than acquiring lock _L2_ to perform intent _i2_ it should first take over _L1_ and perform the intent of _i1_. _i1_ recorded in _L1_ will likely have the same version implied by _i2_ though since worker _B_ will not know about the version, _v1_, intended by _i1_ yet (since it was not committed prior to worker _B_'s attempt), so I think this will always result in object already exists for the first iteration of _i2_.

Am I making sense or have I landed in left field?

**Reply by houqp** [reply]
Ha, I see. You are right. We are on the same page then. The repair action itself will not cause object already exists error, but subsequent version conflict checks would (step 3 and step 4). When a repair happens, we will always get version conflict in the subsequent version conflict checks.

Here is the scenario play:

* worker A tries to commit v1
* worker A acquires L1
* worker A prepares rename commit k1 -> v1
* worker A pauses and L1 expires
* worker B tries to commit v1
* worker B acquires L2
* worker B repair unfinished commit k1 -> v1 (now v1 s3 object is visible to the world)
* worker B gets object already exists error from S3 Heading v1

I am guessing @mosyp was thinking about the optimization of incrementing the version number for append only writes while the lock is being held?

**Reply by xianwill** [reply]
I don't want to speak for @mosyp - but when we were talking this morning, I personally needed a "sit and think" period to work out the implications of the lock placement and was not able to do so in real time. Hazarding a guess, I think letting the optimistic concurrency loop handle the version conflict implied by repair will sit fine with him. How about it @mosyp?

**Reply by houqp** [reply]
If it turned out to be for that optimization, I am also leaning towards doing that as a follow up after we get the reference implementation behavior working end to end first. We likely will only be dealing with single digit dynanmodb calls per second, so there is also the possibility that this will not be a bottleneck for us.

Ideally, the append only version increment optimization should be designed in a way to benefit all storage backends, so it shouldn't depend on a distributed lock.

**Reply by mosyp** [reply]
@xianwill @houqp Yes, having lock outside of rename was purely an optimisation issue. Also I couldn't figure out repairs then, but your logic of returning _version_exists_  and rely on optimistic concurrency works with me. 
Also, I did some testing with lock inside and outside of rename call and couldn't spot much of a difference in https://github.com/delta-io/delta-rs/pull/212. Both cases had random execution time in between 9-10 sec each.

So as of now let's keep dynamo locking within s3_rename until we really need otherwise

**Comment by mosyp** [comment]
@xianwill @houqp wdyt?

## Atomic rename for S3 v2.0

We introduce new `tmp_path` and `log_path` fields into lock's `data` object, which represents an attempt of renaming file `tmp_path` into `log_path`. These fields are added to the lock only within `acquire_lock_with` function.

Variables 
`data1` - data of current lock (if any) that acquired by some worker which in a process of renaming `tmp_path1` into `log_path1`.
`data2` - data of **this** worker that we're trying to rename `tmp_path2` into `log_path2`

The result of `acquire_lock_with(tmp_path2, log_path2)`:

a) The lock is **not** acquired: try again (still in progress or someone else took it).

b) Acquired **released** or **new** lock: 
   1. Acquire lock with `data=tmp_path2/log_path2`
   2. Call `rename(tmp_path2,log_path2)`
   3. Release lock with `data=null` 

c) Current lock is **expired**: this is dangerous zone, previous owner is in undefined state - e.g. halting, crashed etc. At this point we have to repair a rename commit (if any) of previous owner:

1. Acquire lock with the same `data=tmp_path1/log_path1` (because at this point we only repair failed commit, not creating new).
2. Call `rename(tmp_path1,log_path1)` e.g. repair failed rename. 
3. Release lock with `data=null` 
4. Return with `VersionExists` so later **this** worker will have to retry `rename(tmp_path2,log_path2)` for newer version

**Reply by mosyp** [reply]
With this approach we don't need auto incremental `fencing_token` since there's could be only single broken rename attempt at the time, because only one worker can acquire the lock. For now I see only 2 sources of broken renames - worker crashed during rename or worker halted during rename. The dangerous one is latter, as it could wake up and keep worker, but the lock is stolen. However even if so, the rename is idempotent, other worker that acquired the expired lock might already do a rename/repair for us. 

So, taking into account that there's possible only one broken rename commit at the same time, I feel like we don't need incremental token here and might rely on dynamo db `data` field. 

When I'm thinking about that, having this `data` fields in dynamodb is a some sort of that counter, but in another, more specific way.

**Reply by houqp** [reply]
I have been going back and forth on this one as well thinking that we could get away without using a fencing token :P Let me try to provide a possible trace that could break correctness:

Let's say we have 3 writers: A, B, C

0. `lock state after action` action 
1. `data=(tmp_path_a, N)` writer A acquired lock to commit version N with tmp obj `tmp_path_a`
1. `data=(tmp_path_a, N)` writer A completed the rename in S3
1. `data=(tmp_path_a, N)` writer A stalled before it had the chance to release the lock
1. `data=(tmp_path_a, N)` lock expired
1. `data=(tmp_path_a, N)` writer B acquired the lock to commit version N with tmp obj `tmp_path_b`
1. `data=(tmp_path_a, N)` failed conditional PUT notified writer B that it needs to repair `tmp_path_a => _delta_log/N.json` first
1. `data=(tmp_path_a, N)` writer B repaired `tmp_path_a => _delta_log/N.json`
1. `data=null` writer B released lock with `data=null` to mark commit as completed in dynamodb
1. `data=(tmp_path_b, N+1)` writer B acquired the lock to commit version N+1 with data set to `tmp_path_b`
1. `data=(tmp_path_b, N+1)` writer B started the S3 copy
1. `data=(tmp_path_b, N+1)` writer B stalled right before issuing the S3 copy API call
1. `data=(tmp_path_b, N+1)` lock expired, at this point, both writer B and writer A think they own the lock.
1. `data=null` **writer A resumed execution, released the already expired lock** by setting `data=null`
1. `data=(tmp_path_c, N+1)` writer C acquired the lock to commit version N+1 with tmp obj `tmp_path_c`
1. `data=(tmp_path_c, N+1)` writer C completed the S3 copy. now `_delta_log/N+1.json` is visible to all readers with content populated from `tmp_path_c`.
1. `data=null` writer C released the lock.
1. `data=null` writer B resumed and completes its S3 copy by writing `tmp_path_b` to `_delta_log/N+1.json`
1. `data=null` writer B tries to release the lock. The conditional PUT will result in an error, but it's already too late.

At this point, we lost the rename from writer C. This can happen with just two writers as well, but I added writer C to make it easier to follow.

The core of this problem is without using fencing token to signal whether a lock has expired or not, lock owners could incorrectly release locks they don't own anymore. In order words, even though s3 copy is idempotent, the lock release operation is not.

The only way to know whether a lock has expired or not is to associate every lock request with a monotonically increasing counter. Some designs like redis distributed lock uses a unique token to guard the lock, which is not enough to guarantee correctness because the token itself doesn't provide ordering information. The ordering provided by fencing token helps prevent lock owners from incorrectly releasing already expired locks.

**Reply by xianwill** [reply]
Interesting. Makes sense @houqp. One thing that occurs to me is that the intended rename path `_delta_log/\d+.json` could be _used_ in the lock release condition expression as a de facto fencing token since it contains the delta version number which is monotonically increasing. However, this side-effect of the delta log path is specific to our current use case and not generalizable. Seems like we need to keep the fencing token.

**Reply by mosyp** [reply]
I must disagree

> 12. data=(tmp_path_b, N+1) lock expired, at this point, both writer B and writer A think they own the lock.
> 13. data=null **writer A resumed execution, released the already expired lock** by setting data=null

We do have fencing token now which is uuid, which means that worker A won't be able to release a lock.
 
After 12 and 13, we have following situation, worker A fails because it tried to release stolen lock, worker B sleeps.

After steps 14-16 occurred, Workers C's written into N+1, which is ok, then when worker B wakes up and tries to release a lock it'll fail too, since it's gone. 

Basically when there's situation that someone else took its owner lock, then that owner won't be able to release it. We might either kill it's holder or retry, both cases won't influence a delta log.

**Reply by mosyp** [reply]
I'm not against fencing token, this is the key and we already have one (uuid), I just don't see incremental token benefits, since there's only single holder of a lock at the same time. Meaning that only one rename at the time could fail, leading to single repair

**Reply by mosyp** [reply]
What I feel to do is the following, adding repairing on top of the current impl with uuid token will be easier in terms of implementation. Once we have this, then we can compare what we'll have got vs incremental approach. If that won't work for us, we'll change to incremental token. E.g. it doesn't matter much when to implement incremental token, before or after repair logic 

wdyt @xianwill @houqp ?

**Reply by xianwill** [reply]
@mosyp - It makes sense that worker A should not be able to release the lock in the current state which includes the UUID (aka rvn) condition expression, but I am reading the proposal and scenario play as "assume no token/condition-expression at all" (neither monotonic integer nor UUID) rather than current state. Perhaps this was a misread on my part. Sounds like you are suggesting that we stick with the unique token instead of moving to a monotonic token rather than my initial interpretation? Based on steps 12 and 13 - it seems clear that we need a condition expression based on some sort of token to prevent invalid release.

I think this boils down to understanding @houqp 's mention:

> Some designs like redis distributed lock uses a unique token to guard the lock, which is not enough to guarantee correctness because the token itself doesn't provide ordering information.

Specifically understanding how leveraging the ordering semantics of a monotonically increasing fencing token achieves correctness while using a unique non-montonic token does not.

I think @houqp mentioned https://martin.kleppmann.com/2016/02/08/how-to-do-distributed-locking.html before. I'm going to go read that closer right now.

**Reply by xianwill** [reply]
Really great article...  IMHO after reading: I agree with @houqp that a monotonic fencing token is a better overall design that supports a broader range of use cases - which is great for when the time comes to extract the distributed lock into a separate crate, but I also agree with @mosyp that it does not seem to add benefit to our current delta log use case.

The article includes [a section](https://martin.kleppmann.com/2016/02/08/how-to-do-distributed-locking.html#making-the-lock-safe-with-fencing) that describes usage of the fencing token with 2 clients, a lock service and a storage server. Kleppmann mentions:

> Note this requires the storage server to take an active role in checking tokens, and rejecting any writes on which the token has gone backwards.

Basically, the monotonic fencing token is better than a unique token _because_ it _allows_ the storage server to take an active role in rejecting writes from a writer holding an expired lease.

Our "storage server" is really just S3 and is not capable of taking such an active role that would allow it to use the fencing token to reject old tokens. This actually seems fine for our idempotent use case though since the operation performed by a "woke" (previously paused) client will perform exactly the same copy as a repair operation performed on another worker that beats it to the punch. 

So that's my current interpretation. @houqp let me know where you stand or if I missed something important, but I think the decision seems like it should turn on whether we want the distributed lock implementation embedded in delta-rs to support a broader range of use cases that might be able to leverage that active storage server role. In other words, prepping for when we extract it to a separate crate or possible future delta-rs enhancements.

I _think_ what we have now with the conditional check on the unique token on release works fine for our very specific use case that we are dealing with atm.

Both paths forward make sense to me and I'm terrible at making decisions.

**Reply by houqp** [reply]
Ha, yeah, I think both of you are right, I forgot we have a unique owner token to check when releasing the lock to prevent releasing stale locks. Since we don't need to track progress across multiple transaction steps, a unique lock owner identifier is indeed good enough. Let's keep it simple for now and add fencing token later if a real use-case comes up 👍 .

**Reply by mosyp** [reply]
That's great, since we're all on the same page, I feel like the most difficult part is behind us and we good to go :)

**Reply by houqp** [reply]
One thing to watch out for is the (source, dest) pair we store in lock data is not guaranteed to always be valid commit anymore because the conditional put doesn't check for order. The order was enforced in the previous version as a separate conditional PUT step. As a result, in this design we can't always return `already_exists` error after a commit is being "repaired".

For example:

1. writer A wrote and committed Log version N to table
1. writer B, being behind, acquired the lock with data set to `(tmp_b, N)`
1. writer B crashed or paused before releasing the lock
1. writer A acquired the lock with data set to `(tmp_a, N+1)`
1. seeing an existing lock data `(tmp_b, N)`, writer A "repair" the commit and returns `already_exists` error
1. writer A retries with `(tmp_a, N+2)`, which is not correct

So in case of version going backwards, we need to retry lock with the same version within `rename` instead of propagate `already_exists` error back to the optimistic transaction commit loop.

**Reply by xianwill** [reply]
Whew. Wow. Nice catch. I completely missed that scenario. Agree - very important to make sure we satisfy both monotonic _and_ **contiguous** properties of each Delta version number.

From [PROTOCOL.md](https://github.com/delta-io/delta/blob/master/PROTOCOL.md#delta-table-specification):

> A table has a single serial history of atomic versions, which are named using contiguous, monotonically-increasing integers.

**Reply by mosyp** [reply]
Aha, this one https://github.com/delta-io/delta-rs/discussions/89#discussioncomment-67177, so that why the version_exists was chosen. But having the @houqp example described above, we need to definitely change a current design to avoid returning alread_exists.

I agree with 
> So in case of version going backwards, we need to retry lock with the same version within rename instead of propagate already_exists error back to the optimistic transaction commit loop

And, even if repair and current rename will resolve into the same version, the latter retry will fail with already_exists and new version will be picked up by optimistic loop instead of _always_ increasing the version
